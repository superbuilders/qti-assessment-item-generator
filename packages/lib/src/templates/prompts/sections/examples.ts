import { existsSync, readFileSync } from "node:fs"
import * as path from "node:path"
import { resolveLibPath } from "@/internal/paths"

const POSITIVE_EXAMPLE_CACHE = new Map<string, string>()
const NEGATIVE_EXAMPLE_CACHE = new Map<string, string>()

const POSITIVE_ROOT = resolveLibPath("templates/prompts/examples/positive")
const POSITIVE_TEMPLATES_DIR = path.join(POSITIVE_ROOT, "templates")
const POSITIVE_NOTES_DIR = path.join(POSITIVE_ROOT, "notes")

const NEGATIVE_ROOT = resolveLibPath("templates/prompts/examples/negative")
const NEGATIVE_TEMPLATES_DIR = path.join(NEGATIVE_ROOT, "templates")
const NEGATIVE_NOTES_DIR = path.join(NEGATIVE_ROOT, "notes")

export function createPositiveExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadPositiveExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### POSITIVE_EXAMPLES
<examples type="positive">
${rendered}
</examples>`
}

export function createNegativeExampleSection(names: readonly string[]): string {
	if (names.length === 0) return ""
	const rendered = names
		.map((name) => loadNegativeExample(name))
		.filter((block) => block.length > 0)
		.join("\n\n")
	if (rendered.length === 0) return ""
	return `### NEGATIVE_EXAMPLES
<examples type="negative">
${rendered}
</examples>`
}

function loadPositiveExample(name: string): string {
	const cached = POSITIVE_EXAMPLE_CACHE.get(name)
	if (cached) return cached

	const codePath = path.join(POSITIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}

	const notesPath = path.join(POSITIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""

	const rendered = `<example kind="positive" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	POSITIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}

function loadNegativeExample(name: string): string {
	const existing = NEGATIVE_EXAMPLE_CACHE.get(name)
	if (existing) return existing

	const codePath = path.join(NEGATIVE_TEMPLATES_DIR, `${name}.ts`)
	if (!existsSync(codePath)) {
		return ""
	}
	const notesPath = path.join(NEGATIVE_NOTES_DIR, `${name}.md`)
	const code = readFileSync(codePath, "utf-8")
	const notes = existsSync(notesPath)
		? readFileSync(notesPath, "utf-8").trim()
		: ""
	const rendered = `<example kind="negative" name="${name}" source="${codePath}">
<code>
${code}
</code>
${notes ? `<notes>\n${notes}\n</notes>` : "<notes />"}
</example>`
	NEGATIVE_EXAMPLE_CACHE.set(name, rendered)
	return rendered
}
