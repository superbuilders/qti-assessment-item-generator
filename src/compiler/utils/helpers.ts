import { typedSchemas } from "@/widgets/registry"

/**
 * Calculates the greatest common divisor (GCD) of two numbers using the Euclidean algorithm.
 * @param a The first number.
 * @param b The second number.
 * @returns The GCD of a and b.
 */
export function gcd(a: number, b: number): number {
	return b === 0 ? a : gcd(b, a % b)
}

/**
 * Determines if a fraction results in a terminating decimal.
 * A fraction terminates if and only if the prime factors of its simplified denominator are only 2s and 5s.
 * @param numerator The numerator of the fraction.
 * @param denominator The denominator of the fraction.
 * @returns True if the fraction terminates, false otherwise.
 */
export function isTerminatingFraction(
	numerator: number,
	denominator: number
): boolean {
	if (denominator === 0) {
		return false
	}
	// Simplify the fraction by dividing by the GCD.
	const commonDivisor = gcd(Math.abs(numerator), Math.abs(denominator))
	let simplifiedDen = Math.abs(denominator) / commonDivisor

	// The fraction terminates if the simplified denominator's prime factors are only 2 and 5.
	while (simplifiedDen % 2 === 0) {
		simplifiedDen /= 2
	}
	while (simplifiedDen % 5 === 0) {
		simplifiedDen /= 5
	}

	return simplifiedDen === 1
}

export function encodeDataUri(content: string): string {
	const base64 = btoa(unescape(encodeURIComponent(content)))
	const isSvg = content.trim().startsWith("<svg")
	return `${isSvg ? "data:image/svg+xml" : "data:text/html"};base64,${base64}`
}

// Helper function to check if a string is a valid widget type
export function isValidWidgetType(
	type: string
): type is keyof typeof typedSchemas {
	return type in typedSchemas
}

/**
 * Attempts to approximate simple multiplicative/divisive expressions that include π or pi
 * by replacing pi with the provided approximation (default 3.14) and evaluating the expression.
 *
 * Supported forms (case-insensitive for "pi"):
 * - "144*pi", "pi*144", "144pi" (implicit multiplication), "pi144" (rare, but treated as pi*144)
 * - Chains with '*' and '/' only, e.g. "3/2*pi", "pi/4", "6*pi/5", "(144) * (pi)"
 * Parentheses are allowed but only for grouping; no addition/subtraction is supported.
 *
 * Returns a normalized decimal string if evaluation succeeds, otherwise null.
 */
export function tryApproximatePiProduct(
	rawExpr: string,
	piApprox = 3.14
): string | null {
	if (typeof rawExpr !== "string" || rawExpr.trim() === "") return null

	// Quick check: must contain a pi token of some sort
	const containsPi = /\bpi\b/i.test(rawExpr) || rawExpr.includes("π")
	if (!containsPi) return null

	// Normalize: insert explicit * for implicit numeric-pi adjacency in common cases
	const expr = rawExpr
		// insert * between number and pi (e.g., 144pi -> 144*pi)
		.replace(/(\d)\s*(π|\bpi\b)/gi, "$1*$2")
		// and between pi and number (e.g., pi144 -> pi*144)
		.replace(/(π|\bpi\b)\s*(\d)/gi, "$1*$2")
		// remove whitespace
		.replace(/\s+/g, "")
		// drop parentheses (multiplicative-only evaluation does not require them)
		.replace(/[()]/g, "")
		// normalize 'pi' variants to numeric approximation
		.replace(/π/gi, String(piApprox))
		.replace(/\bpi\b/gi, String(piApprox))

	// After substitution, we only allow digits, '.', '*', '/' characters
	if (/[^0-9.*/]/.test(expr)) return null
	// Disallow addition/subtraction in this lightweight evaluator
	if (/[+-]/.test(expr)) return null
	// Must still contain the approximation value to ensure we actually used pi
	if (!expr.includes(String(piApprox))) return null

	// Split into tokens by * and / and evaluate left-to-right (* and / are same precedence)
	const parts: string[] = expr.split(/([*/])/)
	if (parts.length === 0) return null

	// Validate tokens
	for (let i = 0; i < parts.length; i++) {
		const t = parts[i]
		if (i % 2 === 0) {
			// operand
			if (t === "" || Number.isNaN(Number(t))) return null
		} else {
			// operator must be * or /
			if (t !== "*" && t !== "/") return null
		}
	}

	let acc = Number(parts[0])
	for (let i = 1; i < parts.length; i += 2) {
		const op = parts[i]
		const rhs = Number(parts[i + 1])
		if (op === "*") acc = acc * rhs
		else {
			if (rhs === 0) return null
			acc = acc / rhs
		}
	}

	// Normalize number to a compact string while preserving significant digits
	// Avoid scientific notation for typical classroom values
	const normalizeNumberString = () => {
		const maybeFixed = acc.toFixed(10) // limit noise
		// trim trailing zeros
		const trimmed = maybeFixed.replace(/\.0+$/, "").replace(/(\.\d*?)0+$/, "$1")
		return trimmed
	}
	const asString = normalizeNumberString()

	// Add a guard: must be a finite number
	if (!Number.isFinite(Number(asString))) return null
	return asString
}

/**
 * Normalizes a string part for use in a derived QTI identifier.
 * Converts to upper snake case and removes invalid characters.
 * @param part The string part to normalize.
 * @returns The normalized string.
 */
export function normalizeIdPart(part: string): string {
	return part.toUpperCase().replace(/[^A-Z0-9_]/g, "_")
}

/**
 * Derives a deterministic, QTI-safe combo identifier from a path of response parts.
 * @param pathParts An array of strings, where each string is a response-value pair (e.g., "RESPONSE_1_A").
 * @returns The fully-formed combo identifier (e.g., "FB__RESPONSE_1_A__RESPONSE_2_CORRECT").
 */
export function deriveComboIdentifier(pathParts: string[]): string {
	return `FB__${pathParts.join("__")}`
}
