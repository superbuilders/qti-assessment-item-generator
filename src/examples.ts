import type { AssessmentItemInput } from "@/core/item"
import type { allWidgetsCollection } from "@/widgets/collections/all"
import type { WidgetTypeTupleFrom } from "@/widgets/collections/types"

export const allExamples: AssessmentItemInput<WidgetTypeTupleFrom<typeof allWidgetsCollection>>[] = []
