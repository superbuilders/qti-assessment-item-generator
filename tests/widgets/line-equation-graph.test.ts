import { describe, expect, test } from "bun:test"
import * as errors from "@superbuilders/errors"
import * as logger from "@superbuilders/slog"
import { lineEquationGraphExamples } from "../../examples/line-equation-graph"
import type { WidgetInput } from "../../src/widgets/registry"
import { generateWidgetForTest } from "../helpers/generateWidgetForTest"

describe("Widget: line-equation-graph", () => {
	const examples: WidgetInput[] = lineEquationGraphExamples

	examples.forEach((props, index) => {
		test(`should produce consistent output for example #${index + 1}`, async () => {
			const result = await errors.try(generateWidgetForTest(props))
			if (result.error) {
				logger.error("widget generation failed", { error: result.error, index })
				throw result.error
			}
			expect(result.data).toMatchSnapshot()
		})
	})
})
