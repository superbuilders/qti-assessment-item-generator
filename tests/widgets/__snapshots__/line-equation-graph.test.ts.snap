// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`Widget: line-equation-graph should produce consistent output for example #1 1`] = `"<svg width="540" height="882" viewBox="-20 -233 540 882" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="73.4" y="40" width="406.6" height="420"/></clipPath></defs><line x1="73.4" y1="40" x2="73.4" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="93.73" y1="40" x2="93.73" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="114.06" y1="40" x2="114.06" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="134.39000000000001" y1="40" x2="134.39000000000001" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="154.72000000000003" y1="40" x2="154.72000000000003" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="175.05" y1="40" x2="175.05" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="195.38000000000002" y1="40" x2="195.38000000000002" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="215.71" y1="40" x2="215.71" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="236.04000000000002" y1="40" x2="236.04000000000002" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="256.37" y1="40" x2="256.37" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="297.03000000000003" y1="40" x2="297.03000000000003" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="317.36" y1="40" x2="317.36" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="337.69000000000005" y1="40" x2="337.69000000000005" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="358.02" y1="40" x2="358.02" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="378.35" y1="40" x2="378.35" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="398.68000000000006" y1="40" x2="398.68000000000006" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="419.01" y1="40" x2="419.01" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="439.34000000000003" y1="40" x2="439.34000000000003" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="459.6700000000001" y1="40" x2="459.6700000000001" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="480" y1="40" x2="480" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="460" x2="480" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="418" x2="480" y2="418" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="376" x2="480" y2="376" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="334" x2="480" y2="334" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="292" x2="480" y2="292" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="208" x2="480" y2="208" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="166" x2="480" y2="166" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="124" x2="480" y2="124" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="82" x2="480" y2="82" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="40" x2="480" y2="40" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="250" x2="480" y2="250" stroke="#333333" stroke-width="1.5"/><line x1="276.70000000000005" y1="40" x2="276.70000000000005" y2="460" stroke="#333333" stroke-width="1.5"/><line x1="73.4" y1="246" x2="73.4" y2="254" stroke="#333333" stroke-width="1"/><text x="73.4" y="265" font-size="12" text-anchor="middle" fill="#333333">-10</text><line x1="93.73" y1="246" x2="93.73" y2="254" stroke="#333333" stroke-width="1"/><line x1="114.06" y1="246" x2="114.06" y2="254" stroke="#333333" stroke-width="1"/><text x="114.06" y="265" font-size="12" text-anchor="middle" fill="#333333">-8</text><line x1="134.39000000000001" y1="246" x2="134.39000000000001" y2="254" stroke="#333333" stroke-width="1"/><line x1="154.72000000000003" y1="246" x2="154.72000000000003" y2="254" stroke="#333333" stroke-width="1"/><text x="154.72000000000003" y="265" font-size="12" text-anchor="middle" fill="#333333">-6</text><line x1="175.05" y1="246" x2="175.05" y2="254" stroke="#333333" stroke-width="1"/><line x1="195.38000000000002" y1="246" x2="195.38000000000002" y2="254" stroke="#333333" stroke-width="1"/><text x="195.38000000000002" y="265" font-size="12" text-anchor="middle" fill="#333333">-4</text><line x1="215.71" y1="246" x2="215.71" y2="254" stroke="#333333" stroke-width="1"/><line x1="236.04000000000002" y1="246" x2="236.04000000000002" y2="254" stroke="#333333" stroke-width="1"/><text x="236.04000000000002" y="265" font-size="12" text-anchor="middle" fill="#333333">-2</text><line x1="256.37" y1="246" x2="256.37" y2="254" stroke="#333333" stroke-width="1"/><line x1="297.03000000000003" y1="246" x2="297.03000000000003" y2="254" stroke="#333333" stroke-width="1"/><line x1="317.36" y1="246" x2="317.36" y2="254" stroke="#333333" stroke-width="1"/><text x="317.36" y="265" font-size="12" text-anchor="middle" fill="#333333">2</text><line x1="337.69000000000005" y1="246" x2="337.69000000000005" y2="254" stroke="#333333" stroke-width="1"/><line x1="358.02" y1="246" x2="358.02" y2="254" stroke="#333333" stroke-width="1"/><text x="358.02" y="265" font-size="12" text-anchor="middle" fill="#333333">4</text><line x1="378.35" y1="246" x2="378.35" y2="254" stroke="#333333" stroke-width="1"/><line x1="398.68000000000006" y1="246" x2="398.68000000000006" y2="254" stroke="#333333" stroke-width="1"/><text x="398.68000000000006" y="265" font-size="12" text-anchor="middle" fill="#333333">6</text><line x1="419.01" y1="246" x2="419.01" y2="254" stroke="#333333" stroke-width="1"/><line x1="439.34000000000003" y1="246" x2="439.34000000000003" y2="254" stroke="#333333" stroke-width="1"/><text x="439.34000000000003" y="265" font-size="12" text-anchor="middle" fill="#333333">8</text><line x1="459.6700000000001" y1="246" x2="459.6700000000001" y2="254" stroke="#333333" stroke-width="1"/><line x1="480" y1="246" x2="480" y2="254" stroke="#333333" stroke-width="1"/><text x="480" y="265" font-size="12" text-anchor="middle" fill="#333333">10</text><line x1="272.70000000000005" y1="460" x2="280.70000000000005" y2="460" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="464" font-size="12" text-anchor="end" fill="#333333">-10</text><line x1="272.70000000000005" y1="418" x2="280.70000000000005" y2="418" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="422" font-size="12" text-anchor="end" fill="#333333">-8</text><line x1="272.70000000000005" y1="376" x2="280.70000000000005" y2="376" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="380" font-size="12" text-anchor="end" fill="#333333">-6</text><line x1="272.70000000000005" y1="334" x2="280.70000000000005" y2="334" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="338" font-size="12" text-anchor="end" fill="#333333">-4</text><line x1="272.70000000000005" y1="292" x2="280.70000000000005" y2="292" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="296" font-size="12" text-anchor="end" fill="#333333">-2</text><line x1="272.70000000000005" y1="208" x2="280.70000000000005" y2="208" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="212" font-size="12" text-anchor="end" fill="#333333">2</text><line x1="272.70000000000005" y1="166" x2="280.70000000000005" y2="166" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="170" font-size="12" text-anchor="end" fill="#333333">4</text><line x1="272.70000000000005" y1="124" x2="280.70000000000005" y2="124" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="128" font-size="12" text-anchor="end" fill="#333333">6</text><line x1="272.70000000000005" y1="82" x2="280.70000000000005" y2="82" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="86" font-size="12" text-anchor="end" fill="#333333">8</text><line x1="272.70000000000005" y1="40" x2="280.70000000000005" y2="40" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="44" font-size="12" text-anchor="end" fill="#333333">10</text><text x="276.70000000000005" y="495" font-size="14" text-anchor="middle" fill="#333333">r</text><text x="20.000000000000007" y="250" font-size="14" text-anchor="middle" fill="#333333" transform="rotate(-90, 20.000000000000007, 250)"><tspan x="20.000000000000007" dy="0em">y</tspan></text><g clip-path="url(#clip-0)"><line x1="73.4" y1="628" x2="480" y2="-212" stroke="#1fb6ff" stroke-width="2"/></g><text x="175.7468809004186" y="448.6932611934608" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#1fb6ff">p</text></svg>"`;

exports[`Widget: line-equation-graph should produce consistent output for example #2 1`] = `"<svg width="540" height="575" viewBox="-20 -20 540 575" xmlns="http://www.w3.org/2000/svg" font-family="sans-serif" font-size="12"><defs><clipPath id="clip-0"><rect x="73.4" y="40" width="406.6" height="420"/></clipPath></defs><line x1="73.4" y1="40" x2="73.4" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="93.73" y1="40" x2="93.73" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="114.06" y1="40" x2="114.06" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="134.39000000000001" y1="40" x2="134.39000000000001" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="154.72000000000003" y1="40" x2="154.72000000000003" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="175.05" y1="40" x2="175.05" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="195.38000000000002" y1="40" x2="195.38000000000002" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="215.71" y1="40" x2="215.71" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="236.04000000000002" y1="40" x2="236.04000000000002" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="256.37" y1="40" x2="256.37" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="297.03000000000003" y1="40" x2="297.03000000000003" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="317.36" y1="40" x2="317.36" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="337.69000000000005" y1="40" x2="337.69000000000005" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="358.02" y1="40" x2="358.02" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="378.35" y1="40" x2="378.35" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="398.68000000000006" y1="40" x2="398.68000000000006" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="419.01" y1="40" x2="419.01" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="439.34000000000003" y1="40" x2="439.34000000000003" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="459.6700000000001" y1="40" x2="459.6700000000001" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="480" y1="40" x2="480" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="460" x2="480" y2="460" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="418" x2="480" y2="418" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="376" x2="480" y2="376" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="334" x2="480" y2="334" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="292" x2="480" y2="292" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="208" x2="480" y2="208" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="166" x2="480" y2="166" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="124" x2="480" y2="124" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="82" x2="480" y2="82" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="40" x2="480" y2="40" stroke="#e0e0e0" stroke-width="1"/><line x1="73.4" y1="250" x2="480" y2="250" stroke="#333333" stroke-width="1.5"/><line x1="276.70000000000005" y1="40" x2="276.70000000000005" y2="460" stroke="#333333" stroke-width="1.5"/><line x1="73.4" y1="246" x2="73.4" y2="254" stroke="#333333" stroke-width="1"/><text x="73.4" y="265" font-size="12" text-anchor="middle" fill="#333333">-10</text><line x1="93.73" y1="246" x2="93.73" y2="254" stroke="#333333" stroke-width="1"/><line x1="114.06" y1="246" x2="114.06" y2="254" stroke="#333333" stroke-width="1"/><text x="114.06" y="265" font-size="12" text-anchor="middle" fill="#333333">-8</text><line x1="134.39000000000001" y1="246" x2="134.39000000000001" y2="254" stroke="#333333" stroke-width="1"/><line x1="154.72000000000003" y1="246" x2="154.72000000000003" y2="254" stroke="#333333" stroke-width="1"/><text x="154.72000000000003" y="265" font-size="12" text-anchor="middle" fill="#333333">-6</text><line x1="175.05" y1="246" x2="175.05" y2="254" stroke="#333333" stroke-width="1"/><line x1="195.38000000000002" y1="246" x2="195.38000000000002" y2="254" stroke="#333333" stroke-width="1"/><text x="195.38000000000002" y="265" font-size="12" text-anchor="middle" fill="#333333">-4</text><line x1="215.71" y1="246" x2="215.71" y2="254" stroke="#333333" stroke-width="1"/><line x1="236.04000000000002" y1="246" x2="236.04000000000002" y2="254" stroke="#333333" stroke-width="1"/><text x="236.04000000000002" y="265" font-size="12" text-anchor="middle" fill="#333333">-2</text><line x1="256.37" y1="246" x2="256.37" y2="254" stroke="#333333" stroke-width="1"/><line x1="297.03000000000003" y1="246" x2="297.03000000000003" y2="254" stroke="#333333" stroke-width="1"/><line x1="317.36" y1="246" x2="317.36" y2="254" stroke="#333333" stroke-width="1"/><text x="317.36" y="265" font-size="12" text-anchor="middle" fill="#333333">2</text><line x1="337.69000000000005" y1="246" x2="337.69000000000005" y2="254" stroke="#333333" stroke-width="1"/><line x1="358.02" y1="246" x2="358.02" y2="254" stroke="#333333" stroke-width="1"/><text x="358.02" y="265" font-size="12" text-anchor="middle" fill="#333333">4</text><line x1="378.35" y1="246" x2="378.35" y2="254" stroke="#333333" stroke-width="1"/><line x1="398.68000000000006" y1="246" x2="398.68000000000006" y2="254" stroke="#333333" stroke-width="1"/><text x="398.68000000000006" y="265" font-size="12" text-anchor="middle" fill="#333333">6</text><line x1="419.01" y1="246" x2="419.01" y2="254" stroke="#333333" stroke-width="1"/><line x1="439.34000000000003" y1="246" x2="439.34000000000003" y2="254" stroke="#333333" stroke-width="1"/><text x="439.34000000000003" y="265" font-size="12" text-anchor="middle" fill="#333333">8</text><line x1="459.6700000000001" y1="246" x2="459.6700000000001" y2="254" stroke="#333333" stroke-width="1"/><line x1="480" y1="246" x2="480" y2="254" stroke="#333333" stroke-width="1"/><text x="480" y="265" font-size="12" text-anchor="middle" fill="#333333">10</text><line x1="272.70000000000005" y1="460" x2="280.70000000000005" y2="460" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="464" font-size="12" text-anchor="end" fill="#333333">-10</text><line x1="272.70000000000005" y1="418" x2="280.70000000000005" y2="418" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="422" font-size="12" text-anchor="end" fill="#333333">-8</text><line x1="272.70000000000005" y1="376" x2="280.70000000000005" y2="376" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="380" font-size="12" text-anchor="end" fill="#333333">-6</text><line x1="272.70000000000005" y1="334" x2="280.70000000000005" y2="334" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="338" font-size="12" text-anchor="end" fill="#333333">-4</text><line x1="272.70000000000005" y1="292" x2="280.70000000000005" y2="292" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="296" font-size="12" text-anchor="end" fill="#333333">-2</text><line x1="272.70000000000005" y1="208" x2="280.70000000000005" y2="208" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="212" font-size="12" text-anchor="end" fill="#333333">2</text><line x1="272.70000000000005" y1="166" x2="280.70000000000005" y2="166" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="170" font-size="12" text-anchor="end" fill="#333333">4</text><line x1="272.70000000000005" y1="124" x2="280.70000000000005" y2="124" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="128" font-size="12" text-anchor="end" fill="#333333">6</text><line x1="272.70000000000005" y1="82" x2="280.70000000000005" y2="82" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="86" font-size="12" text-anchor="end" fill="#333333">8</text><line x1="272.70000000000005" y1="40" x2="280.70000000000005" y2="40" stroke="#333333" stroke-width="1"/><text x="268.70000000000005" y="44" font-size="12" text-anchor="end" fill="#333333">10</text><text x="276.70000000000005" y="495" font-size="14" text-anchor="middle" fill="#333333">t</text><text x="20.000000000000007" y="250" font-size="14" text-anchor="middle" fill="#333333" transform="rotate(-90, 20.000000000000007, 250)"><tspan x="20.000000000000007" dy="0em">y</tspan></text><g clip-path="url(#clip-0)"><line x1="73.4" y1="533.5" x2="480" y2="218.5" stroke="#14b8a6" stroke-width="2"/></g><text x="204.62486474355802" y="449.5476325769834" font-size="14" text-anchor="middle" dominant-baseline="middle" font-weight="700" fill="#14b8a6">q</text></svg>"`;
